rl4co/envs/routing/cvrptw.py

```diff
@@ -0,0 +1,162 @@
+            low=self.min_time,
+            high=self.max_time,
+            shape=(
+                self.num_loc,
+                2,
+            ),  # each location has a 2D time window (start, end)
+            dtype=torch.float32,
+        )
+
+        num_vehicles = UnboundedDiscreteTensorSpec(shape=(1), dtype=torch.int64)
+        # vehicle_idx = UnboundedDiscreteTensorSpec(shape=(1), dtype=torch.int64)
+
+        # extend observation specs
+        self.observation_spec = CompositeSpec(
+            **self.observation_spec,
+            current_time=current_time,
+            durations=durations,
+            num_vehicles=num_vehicles,
+            time_windows=time_windows,
+            # vehicle_idx=vehicle_idx,
+        )
+
+    def generate_data(self, batch_size) -> TensorDict:
+        td = super().generate_data(batch_size)
+        # batch_size = [td["locs"].shape[0]]
+        batch_size = [batch_size] if isinstance(batch_size, int) else batch_size
+
+        # initialize at zero
+        current_time = torch.zeros(*batch_size, 1, dtype=torch.float32)
+        num_vehicles = torch.ones(*batch_size, dtype=torch.int64)
+
+        # define time windows
+        min_ts = torch.FloatTensor(*batch_size, self.num_loc + 1).uniform_(
+            self.min_time, self.max_time
+        )
+        max_ts = torch.FloatTensor(*batch_size, self.num_loc + 1).uniform_(
+            self.min_time, self.max_time
+        )
+        min_times, max_times = torch.min(min_ts, max_ts), torch.max(min_ts, max_ts)
+        time_windows = torch.stack((min_times, max_times), dim=-1)
+        time_windows[:, 0, :] = 0.0  # depot has no time window
+
+        # first assume service durations of 0 (to be changed later)
+        durations = torch.zeros(*batch_size, self.num_loc + 1, dtype=torch.float32)
+        # for the case later that durations != 0 are used, the durations for the depot must still be 0
+        durations[:, 0] = 0.0
+        td.update(
+            {
+                "current_time": current_time,
+                "durations": durations,
+                "num_vehicles": num_vehicles,
+                "time_windows": time_windows,
+            }
+        )
+        return td
+
+    @staticmethod
+    def get_action_mask(td: TensorDict) -> torch.Tensor:
+        masked = CVRPEnv.get_action_mask(td)
+        current_loc = gather_by_index(td["locs"], td["current_node"])
+        # mask_locs = torch.arange(td["locs"].size(1))[None, :] != td["current_node"]
+        # remaining_locs = (
+        #     td["locs"]
+        #     .masked_select(mask_locs.unsqueeze(-1))
+        #     .reshape([3, td["locs"].size(1) - 1, 2])
+        # )
+        dist = get_distance(current_loc, td["locs"].transpose(0, 1)).transpose(0, 1)
+        can_reach_in_time = (
+            td["current_time"] + td["durations"] + dist
+            # + dist[mask_locs].reshape(td["locs"].size(0), td["locs"].size(1) - 1)
+            <= td["time_windows"][:, :, 1]
```

Comment:
[Minor]

When indexing tensors with multiple initial dimension, it is usually best to use the ellipsis instead (`...`), e.g. `td["time_windows"][..., 1]`
The main reason being clarity and also the fact that any batch size can be used this way, including multiple batch dimensions